File: backend-legacy/src/routes/speedtest.ts
import { Hono } from 'hono';
import { SPEED_TESTS, getResults, deleteResult, runSpeedTests } from '../services/speedtest';

const speedtest = new Hono();

// GET /api/speedtest/tests — public, return available test definitions
speedtest.get('/tests', (c) => {
  // Return test definitions organized by model assignment
  const testList = SPEED_TESTS.map((t) => ({
    id: t.id,
    name: t.name,
    description: t.description,
    category: t.category,
    assignmentKey: t.assignmentKey,
  }));
  return c.json({ tests: testList });
});

// POST /api/speedtest/run — admin only, run the full test suite across all models
speedtest.post('/run', async (c) => {
  if (!c.get('auth').authenticated) {
    return c.json({ error: 'Not authenticated' }, 401);
  }

  try {
    // Run tests across ALL model assignments
    const result = await runSpeedTests();
    return c.json({ result });
  } catch (e) {
    const message = e instanceof Error ? e.message : 'Speed test failed';
    console.error('[speedtest] Run error:', message);
    return c.json({ error: message }, 500);
  }
});

// GET /api/speedtest/results — admin only, get all past results
speedtest.get('/results', async (c) => {
  if (!c.get('auth').authenticated) {
    return c.json({ error: 'Not authenticated' }, 401);
  }

  const results = await getResults();
  return c.json({ results });
});

// DELETE /api/speedtest/results/:id — admin only, delete a result
speedtest.delete('/results/:id', async (c) => {
  if (!c.get('auth').authenticated) {
    return c.json({ error: 'Not authenticated' }, 401);
  }

  const id = c.req.param('id');
  const deleted = await deleteResult(id);
  if (!deleted) {
    return c.json({ error: 'Result not found' }, 404);
  }
  return c.json({ success: true });
});

export default speedtest;

File: backend/app/routes/speedtest.py
"""Speedtest routes — mirrors backend/src/routes/speedtest.ts."""

from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..deps import admin_authenticated
from ..logger import error as log_error
from ..services.speedtest import SPEED_TESTS, delete_result, get_results, run_speed_tests

router = APIRouter()


@router.get("/tests")
async def list_tests() -> dict:
    test_list = [
        {
            "id": t["id"],
            "name": t["name"],
            "description": t["description"],
            "category": t["category"],
            "assignmentKey": t["assignmentKey"],
        }
        for t in SPEED_TESTS
    ]
    return {"tests": test_list}


@router.post("/run")
async def run_tests(request: Request) -> dict:
    if not admin_authenticated(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    try:
        result = await run_speed_tests()
        return {"result": result}
    except Exception as e:  # noqa: BLE001
        msg = str(e) or "Speed test failed"
        log_error("[speedtest] Run error:", msg)
        return JSONResponse({"error": msg}, status_code=500)


@router.get("/results")
async def results(request: Request) -> dict:
    if not admin_authenticated(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    return {"results": await get_results()}


@router.delete("/results/{result_id}")
async def delete_result_route(result_id: str, request: Request) -> dict:
    if not admin_authenticated(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    deleted = await delete_result(result_id)
    if not deleted:
        return JSONResponse({"error": "Result not found"}, status_code=404)
    return {"success": True}